// forge-data-management/src/model/BaseAttributesCreatedUpdatedAttributes.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeDataManagement) {
      root.ForgeDataManagement = {};
    }
    root.ForgeDataManagement.BaseAttributesCreatedUpdatedAttributes = factory(root.ForgeDataManagement.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';




  /**
   * The BaseAttributesCreatedUpdatedAttributes model module.
   * @module model/BaseAttributesCreatedUpdatedAttributes
   */

  /**
   * Constructs a new <code>BaseAttributesCreatedUpdatedAttributes</code>.
   * @alias module:model/BaseAttributesCreatedUpdatedAttributes
   * @class
   * @param createTime {Date} 
   * @param createUserId {String} 
   * @param lastModifiedTime {Date} 
   * @param lastModifiedUserId {String} 
   */
  var exports = function(createTime, createUserId, lastModifiedTime, lastModifiedUserId) {
    var _this = this;

    _this['createTime'] = createTime;
    _this['createUserId'] = createUserId;
    _this['lastModifiedTime'] = lastModifiedTime;
    _this['lastModifiedUserId'] = lastModifiedUserId;
  };

  /**
   * Constructs a <code>BaseAttributesCreatedUpdatedAttributes</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/BaseAttributesCreatedUpdatedAttributes} obj Optional instance to populate.
   * @return {module:model/BaseAttributesCreatedUpdatedAttributes} The populated <code>BaseAttributesCreatedUpdatedAttributes</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('createTime')) {
        obj['createTime'] = ApiClient.convertToType(data['createTime'], 'Date');
      }
      if (data.hasOwnProperty('createUserId')) {
        obj['createUserId'] = ApiClient.convertToType(data['createUserId'], 'String');
      }
      if (data.hasOwnProperty('lastModifiedTime')) {
        obj['lastModifiedTime'] = ApiClient.convertToType(data['lastModifiedTime'], 'Date');
      }
      if (data.hasOwnProperty('lastModifiedUserId')) {
        obj['lastModifiedUserId'] = ApiClient.convertToType(data['lastModifiedUserId'], 'String');
      }
    }
    return obj;
  }

  /**
   * @member {Date} createTime
   */
  exports.prototype['createTime'] = undefined;
  /**
   * @member {String} createUserId
   */
  exports.prototype['createUserId'] = undefined;
  /**
   * @member {Date} lastModifiedTime
   */
  exports.prototype['lastModifiedTime'] = undefined;
  /**
   * @member {String} lastModifiedUserId
   */
  exports.prototype['lastModifiedUserId'] = undefined;



  return exports;
}));
