// forge-data-management/src/model/Hub.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/JsonApiLinks', 'model/JsonApiMeta', 'model/JsonApiResource'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./JsonApiLinks'), require('./JsonApiMeta'), require('./JsonApiResource'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeDataManagement) {
      root.ForgeDataManagement = {};
    }
    root.ForgeDataManagement.Hub = factory(root.ForgeDataManagement.ApiClient, root.ForgeDataManagement.JsonApiLinks, root.ForgeDataManagement.JsonApiMeta, root.ForgeDataManagement.JsonApiResource);
  }
}(this, function(ApiClient, JsonApiLinks, JsonApiMeta, JsonApiResource) {
  'use strict';




  /**
   * The Hub model module.
   * @module model/Hub
   */

  /**
   * Constructs a new <code>Hub</code>.
   * @alias module:model/Hub
   * @class
   * @implements module:model/JsonApiResource
   * @param id {String} resource id
   * @param type {module:model/Hub.TypeEnum} 
   * @param attributes {Object} 
   * @param relationships {Object} 
   */
  var exports = function(id, type, attributes, relationships) {
    var _this = this;

    JsonApiResource.call(_this, id, type);
    _this['type'] = type;
    _this['attributes'] = attributes;
    _this['relationships'] = relationships;
  };

  /**
   * Constructs a <code>Hub</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/Hub} obj Optional instance to populate.
   * @return {module:model/Hub} The populated <code>Hub</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      JsonApiResource.constructFromObject(data, obj);
      if (data.hasOwnProperty('type')) {
        obj['type'] = ApiClient.convertToType(data['type'], 'String');
      }
      if (data.hasOwnProperty('attributes')) {
        obj['attributes'] = ApiClient.convertToType(data['attributes'], Object);
      }
      if (data.hasOwnProperty('relationships')) {
        obj['relationships'] = ApiClient.convertToType(data['relationships'], Object);
      }
    }
    return obj;
  }

  /**
   * @member {module:model/Hub.TypeEnum} type
   */
  exports.prototype['type'] = undefined;
  /**
   * @member {Object} attributes
   */
  exports.prototype['attributes'] = undefined;
  /**
   * @member {Object} relationships
   */
  exports.prototype['relationships'] = undefined;

  // Implement JsonApiResource interface:
  /**
   * resource id
   * @member {String} id
   */
exports.prototype['id'] = undefined;

  /**
   * resource type
   * @member {String} type
   */
exports.prototype['type'] = undefined;

  /**
   * @member {module:model/JsonApiAttributes} attributes
   */
exports.prototype['attributes'] = undefined;

  /**
   * @member {module:model/JsonApiMeta} meta
   */
exports.prototype['meta'] = undefined;

  /**
   * @member {module:model/JsonApiRelationships} relationships
   */
exports.prototype['relationships'] = undefined;

  /**
   * @member {module:model/JsonApiLinks} links
   */
exports.prototype['links'] = undefined;


  /**
   * Allowed values for the <code>type</code> property.
   * @enum {String}
   * @readonly
   */
  exports.TypeEnum = {
    /**
     * value: "hubs"
     * @const
     */
    "hubs": "hubs"  };


  return exports;
}));
