// forge-data-management/src/api/FoldersApi.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/BadInputResponse', 'model/ForbiddenResponse', 'model/FolderResponse', 'model/NotFoundResponse', 'model/JsonApiCollection', 'model/RefsResponse', 'model/CreateRef'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('../model/BadInputResponse'), require('../model/ForbiddenResponse'), require('../model/FolderResponse'), require('../model/NotFoundResponse'), require('../model/JsonApiCollection'), require('../model/RefsResponse'), require('../model/CreateRef'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeDataManagement) {
      root.ForgeDataManagement = {};
    }
    root.ForgeDataManagement.FoldersApi = factory(root.ForgeDataManagement.ApiClient, root.ForgeDataManagement.BadInputResponse, root.ForgeDataManagement.ForbiddenResponse, root.ForgeDataManagement.FolderResponse, root.ForgeDataManagement.NotFoundResponse, root.ForgeDataManagement.JsonApiCollection, root.ForgeDataManagement.RefsResponse, root.ForgeDataManagement.CreateRef);
  }
}(this, function(ApiClient, BadInputResponse, ForbiddenResponse, FolderResponse, NotFoundResponse, JsonApiCollection, RefsResponse, CreateRef) {
  'use strict';

  /**
   * Folders service.
   * @module api/FoldersApi
   */

  /**
   * Constructs a new FoldersApi. 
   * @alias module:api/FoldersApi
   * @class
   * @param {module:ApiClient} apiClient Optional API client implementation to use,
   * default to {@link module:ApiClient#instance} if unspecified.
   */
  var ForgeSDK = require('forge-apis');
  var exports = function(apiClient) {
    this.apiClient = apiClient || ApiClient.instance;

    this.mdClient = new ForgeSDK.FoldersApi();
    this.oauth2_access_code = new ForgeSDK.AuthClientThreeLegged('1234', '1223', 'http://localhost/', ['data:read']);

    /**
     * Callback function to receive the result of the getFolder operation.
     * @callback module:api/FoldersApi~getFolderCallback
     * @param {String} error Error message, if any.
     * @param {module:model/FolderResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the folder by ID for any folder within a given project. All folders or sub-folders within a project are associated with their own unique ID, including the root folder. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {module:api/FoldersApi~getFolderCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/FolderResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/FolderResponse}
     */
    this.getFolderEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}' ;
    this.getFolder = function(projectId, folderId, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getFolder(projectId, folderId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getFolderContents operation.
     * @callback module:api/FoldersApi~getFolderContentsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/JsonApiCollection} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns a collection of items and folders within a folder. Items represent word documents, fusion design files, drawings, spreadsheets, etc. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {Integer} opts.pageNumber specify the page number
     * @param {Integer} opts.pageLimit specify the maximal number of elements per page
     * @param {module:api/FoldersApi~getFolderContentsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/JsonApiCollection}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/JsonApiCollection}
     */
    this.getFolderContentsEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}/contents' ;
    this.getFolderContents = function(projectId, folderId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getFolderContents(projectId, folderId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getFolderParent operation.
     * @callback module:api/FoldersApi~getFolderParentCallback
     * @param {String} error Error message, if any.
     * @param {module:model/FolderResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the parent folder (if it exists). In a project, subfolders and resource items are stored under a folder except the root folder which does not have a parent of its own. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {module:api/FoldersApi~getFolderParentCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/FolderResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/FolderResponse}
     */
    this.getFolderParentEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}/parent' ;
    this.getFolderParent = function(projectId, folderId, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getFolderParent(projectId, folderId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getFolderRefs operation.
     * @callback module:api/FoldersApi~getFolderRefsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/JsonApiCollection} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the resources (&#x60;items&#x60;, &#x60;folders&#x60;, and &#x60;versions&#x60;) which have a custom relationship with the given &#x60;folder_id&#x60;. Custom relationships can be established between a folder and other resources within the &#39;data&#39; domain service (folders, items, and versions). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {module:api/FoldersApi~getFolderRefsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/JsonApiCollection}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/JsonApiCollection}
     */
    this.getFolderRefsEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}/refs' ;
    this.getFolderRefs = function(projectId, folderId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getFolderRefs(projectId, folderId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getFolderRelationshipsRefs operation.
     * @callback module:api/FoldersApi~getFolderRelationshipsRefsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/RefsResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the custom relationships that are associated to the given &#x60;folder_id&#x60;. Custom relationships can be established between a folder and other resources within the &#39;data&#39; domain service (folders, items, and versions). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterRefType filter by &#x60;refType&#x60;
     * @param {module:model/String} opts.filterDirection filter by the direction of the reference
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {module:api/FoldersApi~getFolderRelationshipsRefsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/RefsResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RefsResponse}
     */
    this.getFolderRelationshipsRefsEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}/relationships/refs' ;
    this.getFolderRelationshipsRefs = function(projectId, folderId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getFolderRelationshipsRefs(projectId, folderId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the postFolderRelationshipsRef operation.
     * @callback module:api/FoldersApi~postFolderRelationshipsRefCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Creates a custom relationship between a folder and another resource within the &#39;data&#39; domain service (folder, item, or version). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} folderId the &#x60;folder id&#x60;
     * @param {module:model/CreateRef} body describe the ref to be created
     * @param {module:api/FoldersApi~postFolderRelationshipsRefCallback} callback The callback function, accepting three arguments: error, data, response
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}
     */
    this.postFolderRelationshipsRefEndPoint ='/data/v1/projects/{project_id}/folders/{folder_id}/relationships/refs' ;
    this.postFolderRelationshipsRef = function(projectId, folderId, body, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.postFolderRelationshipsRef(projectId, folderId, body, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };
  };

  return exports;
}));
