// forge-data-management/src/api/ItemsApi.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/HealthResponse', 'model/ItemResponse', 'model/BadInputResponse', 'model/ForbiddenResponse', 'model/NotFoundResponse', 'model/FolderResponse', 'model/JsonApiCollection', 'model/RefsResponse', 'model/VersionResponse', 'model/VersionsResponse', 'model/CreateRef'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('../model/HealthResponse'), require('../model/ItemResponse'), require('../model/BadInputResponse'), require('../model/ForbiddenResponse'), require('../model/NotFoundResponse'), require('../model/FolderResponse'), require('../model/JsonApiCollection'), require('../model/RefsResponse'), require('../model/VersionResponse'), require('../model/VersionsResponse'), require('../model/CreateRef'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeDataManagement) {
      root.ForgeDataManagement = {};
    }
    root.ForgeDataManagement.ItemsApi = factory(root.ForgeDataManagement.ApiClient, root.ForgeDataManagement.HealthResponse, root.ForgeDataManagement.ItemResponse, root.ForgeDataManagement.BadInputResponse, root.ForgeDataManagement.ForbiddenResponse, root.ForgeDataManagement.NotFoundResponse, root.ForgeDataManagement.FolderResponse, root.ForgeDataManagement.JsonApiCollection, root.ForgeDataManagement.RefsResponse, root.ForgeDataManagement.VersionResponse, root.ForgeDataManagement.VersionsResponse, root.ForgeDataManagement.CreateRef);
  }
}(this, function(ApiClient, HealthResponse, ItemResponse, BadInputResponse, ForbiddenResponse, NotFoundResponse, FolderResponse, JsonApiCollection, RefsResponse, VersionResponse, VersionsResponse, CreateRef) {
  'use strict';

  /**
   * Items service.
   * @module api/ItemsApi
   */

  /**
   * Constructs a new ItemsApi. 
   * @alias module:api/ItemsApi
   * @class
   * @param {module:ApiClient} apiClient Optional API client implementation to use,
   * default to {@link module:ApiClient#instance} if unspecified.
   */
  var ForgeSDK = require('forge-apis');
  var exports = function(apiClient) {
    this.apiClient = apiClient || ApiClient.instance;

    this.mdClient = new ForgeSDK.ItemsApi();
    this.oauth2_access_code = new ForgeSDK.AuthClientThreeLegged('1234', '1223', 'http://localhost/', ['data:read']);

    /**
     * Callback function to receive the result of the getDataHealth operation.
     * @callback module:api/ItemsApi~getDataHealthCallback
     * @param {String} error Error message, if any.
     * @param {module:model/HealthResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the health information for the &#x60;data&#x60; domain. The health check assesses the current working status of the &#39;data&#39; domain service. 
     * @param {module:api/ItemsApi~getDataHealthCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/HealthResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/HealthResponse}
     */
    this.getDataHealthEndPoint ='/data/v1/health' ;
    this.getDataHealth = function(callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getDataHealth(oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItem operation.
     * @callback module:api/ItemsApi~getItemCallback
     * @param {String} error Error message, if any.
     * @param {module:model/ItemResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns a resource item by ID for any item within a given project. Resource items represent word documents, fusion design files, drawings, spreadsheets, etc. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {module:api/ItemsApi~getItemCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/ItemResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/ItemResponse}
     */
    this.getItemEndPoint ='/data/v1/projects/{project_id}/items/{item_id}' ;
    this.getItem = function(projectId, itemId, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItem(pprojectId, itemId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItemParentFolder operation.
     * @callback module:api/ItemsApi~getItemParentFolderCallback
     * @param {String} error Error message, if any.
     * @param {module:model/FolderResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the \&quot;parent\&quot; folder for the given item. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {module:api/ItemsApi~getItemParentFolderCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/FolderResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/FolderResponse}
     */
    this.getItemParentFolderEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/parent' ;
    this.getItemParentFolder = function(projectId, itemId, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItemParentFolder(projectId, itemId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItemRefs operation.
     * @callback module:api/ItemsApi~getItemRefsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/JsonApiCollection} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the resources (&#x60;items&#x60;, &#x60;folders&#x60;, and &#x60;versions&#x60;) which have a custom relationship with the given &#x60;item_id&#x60;. Custom relationships can be established between an item and other resources within the &#39;data&#39; domain service (folders, items, and versions). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {module:api/ItemsApi~getItemRefsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/JsonApiCollection}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/JsonApiCollection}
     */
    this.getItemRefsEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/refs' ;
    this.getItemRefs = function(projectId, itemId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItemRefs(projectId, itemId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItemRelationshipsRefs operation.
     * @callback module:api/ItemsApi~getItemRelationshipsRefsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/RefsResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the custom relationships that are associated to the given &#x60;item_id&#x60;. Custom relationships can be established between an item and other resources within the &#39;data&#39; domain service (folders, items, and versions). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterRefType filter by &#x60;refType&#x60;
     * @param {module:model/String} opts.filterDirection filter by the direction of the reference
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {module:api/ItemsApi~getItemRelationshipsRefsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/RefsResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/RefsResponse}
     */
    this.getItemRelationshipsRefsEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/relationships/refs' ;
    this.getItemRelationshipsRefs = function(projectId, itemId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItemRelationshipsRefs(projectId, itemId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItemTip operation.
     * @callback module:api/ItemsApi~getItemTipCallback
     * @param {String} error Error message, if any.
     * @param {module:model/VersionResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns the \&quot;tip\&quot; version for the given item. Multiple versions of a resource item can be uploaded in a project. The tip version is the most recent one. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {module:api/ItemsApi~getItemTipCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/VersionResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/VersionResponse}
     */
    this.getItemTipEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/tip' ;
    this.getItemTip = function(projectId, itemId, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItemTip(projectId, itemId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the getItemVersions operation.
     * @callback module:api/ItemsApi~getItemVersionsCallback
     * @param {String} error Error message, if any.
     * @param {module:model/VersionsResponse} data The data returned by the service call.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Returns versions for the given item. Multiple versions of a resource item can be uploaded in a project. 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {Object} opts Optional parameters
     * @param {Array.<String>} opts.filterType filter by the &#x60;type&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterId filter by the &#x60;id&#x60; of the &#x60;ref&#x60; target
     * @param {Array.<String>} opts.filterExtensionType filter by the extension type
     * @param {Array.<Integer>} opts.filterVersionNumber filter by &#x60;versionNumber&#x60;
     * @param {Integer} opts.pageNumber specify the page number
     * @param {Integer} opts.pageLimit specify the maximal number of elements per page
     * @param {module:api/ItemsApi~getItemVersionsCallback} callback The callback function, accepting three arguments: error, data, response
     * data is of type: {@link module:model/VersionsResponse}
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}, with data of type {@link module:model/VersionsResponse}
     */
    this.getItemVersionsEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/versions' ;
    this.getItemVersions = function(projectId, itemId, opts, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.getItemVersions(projectId, itemId, opts, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };

    /**
     * Callback function to receive the result of the postItemRelationshipsRef operation.
     * @callback module:api/ItemsApi~postItemRelationshipsRefCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Creates a custom relationship between an item and another resource within the &#39;data&#39; domain service (folder, item, or version). 
     * @param {String} projectId the &#x60;project id&#x60;
     * @param {String} itemId the &#x60;item id&#x60;
     * @param {module:model/CreateRef} body describe the ref to be created
     * @param {module:api/ItemsApi~postItemRelationshipsRefCallback} callback The callback function, accepting three arguments: error, data, response
     * @return {Promise} a {@link https://www.promisejs.org/|Promise}
     */
    this.postItemRelationshipsRefEndPoint ='/data/v1/projects/{project_id}/items/{item_id}/relationships/refs' ;
    this.postItemRelationshipsRef = function(projectId, itemId, body, callback) {
			//this.apiClient.applyToAuthObjects (this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']) ;
			var oauth = this.apiClient.getAuthObject(this.oauth2_access_code, this.oauth2_application, ['oauth2_access_code']);
			var pr = this.mdClient.postItemRelationshipsRef(pprojectId, itemId, oauth, oauth.credentials);
			if (callback === undefined) {
				return (new Promise(function (resolve, reject) {
					pr.then(function (result) { resolve(result.body); })
						.catch(function (err) { reject(err); });
				})) ;
			}

			pr.then(function (result) { callback(null, result.body); })
				.catch(function (err) { callback(err, null); });
    };
  };

  return exports;
}));
