// forge-data-management/src/model/JsonApiResource.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'model/JsonApiAttributes', 'model/JsonApiLinks', 'model/JsonApiMeta', 'model/JsonApiRelationships', 'model/JsonApiTypeId'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./JsonApiAttributes'), require('./JsonApiLinks'), require('./JsonApiMeta'), require('./JsonApiRelationships'), require('./JsonApiTypeId'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeDataManagement) {
      root.ForgeDataManagement = {};
    }
    root.ForgeDataManagement.JsonApiResource = factory(root.ForgeDataManagement.ApiClient, root.ForgeDataManagement.JsonApiAttributes, root.ForgeDataManagement.JsonApiLinks, root.ForgeDataManagement.JsonApiMeta, root.ForgeDataManagement.JsonApiRelationships, root.ForgeDataManagement.JsonApiTypeId);
  }
}(this, function(ApiClient, JsonApiAttributes, JsonApiLinks, JsonApiMeta, JsonApiRelationships, JsonApiTypeId) {
  'use strict';




  /**
   * The JsonApiResource model module.
   * @module model/JsonApiResource
   */

  /**
   * Constructs a new <code>JsonApiResource</code>.
   * @alias module:model/JsonApiResource
   * @class
   * @implements module:model/JsonApiTypeId
   * @param id {String} resource id
   * @param type {String} resource type
   */
  var exports = function(id, type) {
    var _this = this;

    JsonApiTypeId.call(_this, id, type);




  };

  /**
   * Constructs a <code>JsonApiResource</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/JsonApiResource} obj Optional instance to populate.
   * @return {module:model/JsonApiResource} The populated <code>JsonApiResource</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      JsonApiTypeId.constructFromObject(data, obj);
      if (data.hasOwnProperty('attributes')) {
        obj['attributes'] = JsonApiAttributes.constructFromObject(data['attributes']);
      }
      if (data.hasOwnProperty('meta')) {
        obj['meta'] = JsonApiMeta.constructFromObject(data['meta']);
      }
      if (data.hasOwnProperty('relationships')) {
        obj['relationships'] = JsonApiRelationships.constructFromObject(data['relationships']);
      }
      if (data.hasOwnProperty('links')) {
        obj['links'] = JsonApiLinks.constructFromObject(data['links']);
      }
    }
    return obj;
  }

  /**
   * @member {module:model/JsonApiAttributes} attributes
   */
  exports.prototype['attributes'] = undefined;
  /**
   * @member {module:model/JsonApiMeta} meta
   */
  exports.prototype['meta'] = undefined;
  /**
   * @member {module:model/JsonApiRelationships} relationships
   */
  exports.prototype['relationships'] = undefined;
  /**
   * @member {module:model/JsonApiLinks} links
   */
  exports.prototype['links'] = undefined;

  // Implement JsonApiTypeId interface:
  /**
   * resource id
   * @member {String} id
   */
exports.prototype['id'] = undefined;

  /**
   * resource type
   * @member {String} type
   */
exports.prototype['type'] = undefined;



  return exports;
}));
